### ISSUE TITLE
likely arm64 miscompile

### ISSUE THREAD

**Issue by regehr** [llvm/llvm-project#55201]
ok this has been llvm-reduced but it's still a bit complex, sorry!

```llvm
define i32 @f(i32 %x) {
  %or1 = or i32 %x, 1
  %sh1 = shl i32 %or1, 5
  %sh2 = lshr i32 %x, 27
  %1 = and i32 %sh2, 1
  %r = or i32 %sh1, %1
  ret i32 %r
}
```

we believe that `f(0xf0000000) -> 32` and that is what we get when using global isel, and also when compiling this code for x64.

on the other hand, the default arm64 backend from top of tree gives `f(0xf0000000) -> 62`

we don't think there's any UB going on here, seems like a straight-up miscompile

cc @ornata @nunoplopes @ryan-berger @nbushehri @aqjune @hatsunespica

**Comment by regehr** [comment]
bad asm from the default backend 
```
_f:
	ror	w8, w0, #27
	orr	w0, w8, #0x20
	ret
```

good asm from global isel:
```
_f: 
	orr	w8, w0, #0x1
	ubfx	w9, w0, #27, #1
	orr	w0, w9, w8, lsl #5
	ret

```

**Comment by topperc** [comment]
-mtriple=riscv32 -mattr=+zbb produces something equivalent to the bad AArch64 code.

**Comment by topperc** [comment]
Seems to be something in the rotate matching. x86 picks a funnel shift instead.

**Comment by topperc** [comment]
This patch fixes it

```
diff --git a/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp b/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
index 71618eb2bd7c..ac9670746c53 100644
--- a/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
+++ b/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
@@ -7482,7 +7482,7 @@ SDValue DAGCombiner::MatchRotate(SDValue LHS, SDValue RHS, const SDLoc &DL) {
 
   // TODO: Support pre-legalization funnel-shift by constant.
   bool IsRotate = LHSShift.getOperand(0) == RHSShift.getOperand(0);
-  if (!IsRotate && !(HasFSHL || HasFSHR)) {
+  if (!IsRotate && !(HasFSHL || HasFSHR) && !LHSMask && !RHSMask) {
     if (TLI.isTypeLegal(VT) && LHS.hasOneUse() && RHS.hasOneUse() &&
         ISD::matchBinaryPredicate(LHSShiftAmt, RHSShiftAmt, MatchRotateSum)) {
       // Look for a disguised rotate by constant.
```

We might be able to do better. The block of code behind that `if` forgot to do anything with LHSMask and RHSMask so this patch makes it so we don't enter the code when we need to do something with them.

**Comment by topperc** [comment]
Candidate patch https://reviews.llvm.org/D124711

**Comment by topperc** [comment]
Fixed by 6affe87bda203b6d6538f41bb44879509962c695
